File: src/publishers/notion/NotionBriefingPublisher.ts
import type { BriefingPublisher } from "../BriefingPublisher.js";
import type {
  PublishBriefingRequest,
  PublishBriefingResult,
  PublishChannelResult,
} from "../../domain/briefing.js";
import { ErrorCodes } from "../../errors/errorCodes.js";
import { nowISOStringKST } from "../../utils/date.js";
import { buildNotionBriefingBlocks } from "./buildNotionPage.js";
import {
  getSafeNotionErrorMessage,
  isRetryableNotionError,
} from "./NotionClientAdapter.js";
import type { NotionBriefingPageClient } from "./notionTypes.js";

export interface NotionBriefingPublisherOptions {
  databaseId: string;
  client: NotionBriefingPageClient;
}

export class NotionBriefingPublisher implements BriefingPublisher {
  private readonly databaseId: string;
  private readonly client: NotionBriefingPageClient;

  constructor(options: NotionBriefingPublisherOptions) {
    this.databaseId = options.databaseId;
    this.client = options.client;
  }

  async publish(request: PublishBriefingRequest): Promise<PublishBriefingResult> {
    const channelResult = await this.publishToNotion(request);

    return {
      briefingId: request.briefing.id,
      results: [channelResult],
      completedAt: nowISOStringKST(),
    };
  }

  private async publishToNotion(
    request: PublishBriefingRequest,
  ): Promise<PublishChannelResult> {
    if (request.dryRun) {
      return {
        channel: "notion",
        status: "skipped",
        externalId: `dry-run:${request.briefing.id}`,
      };
    }

    try {
      const existingPage = await this.client.findPageByBriefingId(
        this.databaseId,
        request.briefing.id,
      );

      if (existingPage !== undefined) {
        return {
          channel: "notion",
          status: "skipped",
          externalId: existingPage.id,
          errorCode: ErrorCodes.PUBLISH_DUPLICATE,
          errorMessage: "이미 같은 브리핑 ID로 저장된 Notion 페이지가 있습니다.",
        };
      }

      const page = await this.client.createBriefingPage({
        databaseId: this.databaseId,
        briefing: request.briefing,
        children: buildNotionBriefingBlocks(request.briefing),
      });

      return {
        channel: "notion",
        status: "success",
        externalId: page.id,
      };
    } catch (error) {
      return {
        channel: "notion",
        status: "failed",
        errorCode: ErrorCodes.PUBLISH_CHANNEL_ERROR,
        errorMessage: getSafeNotionErrorMessage(error),
      };
    }
  }

  static isRetryableFailure(error: unknown): boolean {
    return isRetryableNotionError(error);
  }
}

File: src/main/java/com/economicbriefing/publisher/notion/NotionBriefingPublisher.java
package com.economicbriefing.publisher.notion;

import java.util.List;

import com.economicbriefing.config.NotionProperties;
import com.economicbriefing.exception.ErrorCode;
import com.economicbriefing.publisher.BriefingPublisher;
import com.economicbriefing.publisher.dto.PublishBriefingRequest;
import com.economicbriefing.publisher.dto.PublishBriefingResult;
import com.economicbriefing.publisher.dto.PublishChannelResult;
import com.economicbriefing.util.KstDateTimeUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Service;

@Service
@ConditionalOnProperty(name = "briefing.dry-run", havingValue = "false")
public class NotionBriefingPublisher implements BriefingPublisher {

    private static final Logger log = LoggerFactory.getLogger(NotionBriefingPublisher.class);

    private final NotionClient notionClient;
    private final NotionProperties notionProperties;

    public NotionBriefingPublisher(NotionClient notionClient, NotionProperties notionProperties) {
        this.notionClient = notionClient;
        this.notionProperties = notionProperties;
    }

    @Override
    public PublishBriefingResult publish(PublishBriefingRequest request) {
        PublishChannelResult channelResult = publishToNotion(request);

        return new PublishBriefingResult(
                request.briefing().id(),
                List.of(channelResult),
                KstDateTimeUtil.now()
        );
    }

    private PublishChannelResult publishToNotion(PublishBriefingRequest request) {
        if (request.dryRun()) {
            log.info("Dry-run mode: skipping Notion publish for briefing={}", request.briefing().id());
            return PublishChannelResult.skipped("notion", "dry-run:" + request.briefing().id());
        }

        try {
            // Check for duplicates
            String existingPageId = notionClient.findPageByBriefingId(
                    notionProperties.databaseId(), request.briefing().id());

            if (existingPageId != null) {
                log.warn("Duplicate briefing found in Notion: pageId={}, briefingId={}",
                        existingPageId, request.briefing().id());
                return PublishChannelResult.skipped(
                        "notion",
                        existingPageId,
                        ErrorCode.PUBLISH_DUPLICATE.name(),
                        "이미 같은 브리핑 ID로 저장된 Notion 페이지가 있습니다."
                );
            }

            // Build blocks and create page
            List<NotionBlock> blocks = NotionPageBuilder.buildBriefingBlocks(request.briefing());
            String pageId = notionClient.createBriefingPage(
                    notionProperties.databaseId(), request.briefing(), blocks);

            log.info("Published briefing to Notion: pageId={}, briefingId={}",
                    pageId, request.briefing().id());
            return PublishChannelResult.success("notion", pageId);

        } catch (Exception e) {
            log.error("Failed to publish briefing to Notion: briefingId={}",
                    request.briefing().id(), e);
            return PublishChannelResult.failed(
                    "notion",
                    ErrorCode.PUBLISH_CHANNEL_ERROR.name(),
                    e.getMessage() != null ? e.getMessage() : "Notion 저장 중 알 수 없는 오류가 발생했습니다."
            );
        }
    }
}
